### ISSUE TITLE
Some questions about the paper

### ISSUE THREAD

**Issue by Vtecbest** [siavashzk/BranchNet#3]
![image](https://user-images.githubusercontent.com/72300677/128509142-ce56af0d-2fe9-460d-bce3-b49806df5a63.png)
Hi, may I know what the weight table stands for in this pic? Is it the weights of the FC layer? Also, I don't understand why there are 41 branch slices, I thought the input was global history so it was the same for every branch.

**Comment by siavashzk** [comment]
Hi,

- Yes the weight table contains the weight of FC layers only.
- There are 41 "sets" of slices. Mini-BranchNet iso-latency configuration in the paper supports predicting up to 41 static branches. So we need 41 sets of convolution tables, sum-pooling, buffers, etc. to maintain the history for all 41.
- Conceptually, the input is the global history for every branch. However, doing all the computation necessary for parsing a 600-branch global history in a few cycles is expensive and slow. Optimization 1 in the paper explains how we avoid this by speculatively maintaining convolutional histories for each branch.

**Comment by Vtecbest** [comment]
So you mean for different branches, they share the same hash input but don't share the same conv tables etc., is this right?

**Comment by Vtecbest** [comment]
For me, the most confused part is that I can't figure out the difference between these 41 "sets" of branches. I think at least for slice with the same history length, even for different branch id they should have the same conv table. I'll appreciate it if you can answer this.

**Comment by siavashzk** [comment]
Let me explain the a more general view of the inference engine, maybe it helps answer your questions.

Each static branch is trained completely independently and thus has its own trained model. Therefore, for each branch, the inference engine should first gather the weights for a branch that it wants to predict, and then compute the prediction. So, you need to think about the inference engine from two points of view:
1- How do you get the model weights for the specific branch you're predicting?
2- Given one model for one branch, how do you compute the outcomes of the CNN?

The more branches we want to predict using BranchNet, the more storage we need on the chip to store all the models. The number 41 is the maximum capacity of the iso-latency Mini-BranchNet evaluation. Other configurations (e.g. iso-storage support even fewer number of branches).

Conv table is just a transformation of the embedding+conv+batch norm+binarized sigmoid. Since we train an independent model for each static branch, we end up with different weights, and therefore different conv tables for each branch. Makes sense?

**Comment by siavashzk** [comment]
Note that using independent models for each branch is a design choice. In theory, you could train one big model that predicts all branches. I have not quantitatively compared the two approaches recently, so I could be wrong, but I think training many independent models results in a better accuracy/design complexity trade-off.

**Comment by Vtecbest** [comment]
I see, the idea is brilliant. So every time you want to run a program, you need to profile it first, then train the highest MPKI branches so that you can get the model weight, right?

**Comment by siavashzk** [comment]
Correct!

**Comment by chnjstyj** [comment]
Can I ask a few more questions about the same picture?
What is the mean of the update pipeline inputs "Hash of 3 most recent branches". It is the address of the branches or something else?
Is the weight table contains the offline training content? The weights we get in the offline training steps were stored in the weight table served for online predictions in a real cpu's predictor?

**Comment by siavashzk** [comment]
This is a hash of the <pc,dir> pairs of the 3 youngest branches in the global history.  See hash_using_metadata:
https://github.com/siavashzk/BranchNet/blob/43c1c02e6275f691f880f795ed28f9ebc53b4366/src/branchnet/model.py#L170

Yeah, the weight table and the lookup tables representing the approximation of the convolution are trained offline, and there needs to be a mechanism to load these training data on the chip to be used by the branch predictor at runtime.

**Comment by chnjstyj** [comment]
So is that meaning to Mini-BranchNet 0.25KB config there exists 3 (number of History sizes) * 41 (number of branches to predict) Conv tables?
